ipq40xx: add support for EZVIZ CS-W3-WD1200G EUP

Tom-Brouwer opened the pull request:
#### Hardware:
```
SOC:    Qualcomm IPQ4018
RAM:	128 MB Nanya NT5CC64M16GP-DI
FLASH:  16 MB Macronix MX25L12805D
ETH:    Qualcomm QCA8075 (4 Gigabit ports, 3xLAN, 1xWAN)
WLAN:   Qualcomm IPQ4018 (2.4 & 5 Ghz)
BUTTON: Shared WPS/Reset button
LED:    RGB Status/Power LED
SERIAL: Header J8 (UART, Left side of board). Numbered from
        top to bottom:
        (1) GND, (2) TX, (3) RX, (4) VCC (White triangle
        next to it).
        3.3v, 115200, 8N1
```

#### Tested/Working:
* Ethernet
* WiFi (2.4 and 5GHz)
* Status LED
* Reset Button (See note below)

#### Implementation notes:
* The shared WPS/Reset button is implemented as a Reset button
* I could not find a original firmware image to reverse engineer, meaning
currently it's not possible to flash OpenWRT through the Web GUI.

#### Installation (Through Serial console & TFTP):
1. Set your PC to fixed IP 192.168.1.12, Netmask 255.255.255.0, and connect to
one of the LAN ports
2. Rename the initramfs image to 'C0A8010B.img' and enable a TFTP server on
your pc, to serve the image
2. Connect to the router through serial (See connection properties above)
3. Hit a key during startup, to pause startup
4. type `setenv serverip 192.168.1.12`, to set the tftp server address
5. type `tftpboot`, to load the image from the laptop through tftp
6. type `bootm` to run the loaded image from memory
6. (If you want to return to stock firmware later, create an full MTD backup,
e.g. using instructions here https://openwrt.org/docs/guide-user/installation/generic.backup#create_full_mtd_backup)
7. Transfer the 'sysupgrade' OpenWRT firmware image from PC to router, e.g.:
`scp xxx-squashfs-sysupgrade.bin root@192.168.1.1:/tmp/upgrade.bin`
8. Run sysupgrade to permanently install OpenWRT to flash: `sysupgrade -n /tmp/upgrade.bin`

#### Revert to stock:
To revert to stock, you need the MTD backup from step 6 above:
1. Unpack the MTD backup archive
2. Transfer the 'firmware' partition image to the router (e.g. mtd8_firmware.backup)
3. On the router, do `mtd write mtd8_firmware.backup firmware`

### Note about device name
I've decided to use the full type number for the Model name in OpenWRT. Although the router's short name is call EZVIZ W3, the FCC application seems to suggest there are many different versions (https://fccid.io/2APV2-W3/Label/Label-3966701)

adrianschmutzler commented:
> LAN/WAN Addresses are reversed w.r.t factory firmware ON PURPOSE. On factory
> firmware the MAC Address on the device sticker is (erroneously?) equal to the
> LAN MAC address rather than the WAN MAC address.

Please assign MAC addresses as in stock firmware. Vendors assign the label MAC address relatively randomly to one of the interfaces anyway.

adrianschmutzler commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -27,7 +27,8 @@ ALLWIFIBOARDS:= \
 	linksys_ea8300 \
-	qxwlan_e2600ac
+	qxwlan_e2600ac \
+	ezviz_cs-w3-wd1200g-eup
```
Please use correct alphabetic sorting for positioning, same goes for several cases below.

adrianschmutzler commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -94,5 +95,6 @@ $(eval $(call generate-ipq-wifi-package,aruba_ap-303,Aruba AP-303))
 $(eval $(call generate-ipq-wifi-package,avm_fritzrepeater-1200,AVM FRITZRepeater 1200))
 $(eval $(call generate-ipq-wifi-package,linksys_ea8300,Linksys EA8300))
 $(eval $(call generate-ipq-wifi-package,qxwlan_e2600ac,Qxwlan E2600AC))
+$(eval $(call generate-ipq-wifi-package,ezviz_cs-w3-wd1200g-eup,EZVIZ CS-W3-WD1200G EUP))
```
Sorting.

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -70,6 +70,11 @@ ipq40xx_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0u@eth0" "1:lan" "2:lan" "3:lan" "4:lan" "0u@eth1" "5:wan"
 		;;
+	ezviz,cs-w3-wd1200g-eup)
```
Sorting.

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -108,6 +108,10 @@ case "$FIRMWARE" in
 		caldata_extract "ART" 0x1000 0x2f20
 		ath10k_patch_mac $(macaddr_add $(cat /sys/class/net/eth0/address) -2)
 		;;
+	ezviz,cs-w3-wd1200g-eup)
```
Sorting.

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -177,6 +181,10 @@ case "$FIRMWARE" in
 		caldata_extract "ART" 0x5000 0x2f20
 		ath10k_patch_mac $(macaddr_add $(cat /sys/class/net/eth0/address) -1)
 		;;
+	ezviz,cs-w3-wd1200g-eup)
```
Sorting.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-cs-w3-wd1200g-eup.dts:
```diff
@@ -0,0 +1,258 @@
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
```
Please use a proper SPDX-License-Identifier instead.

Despite, please fix the indent in this entire file to use tabs consistently.

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -500,4 +500,20 @@ define Device/zyxel_wre6606
+	SOC := qcom-ipq4018
+	IMAGES := sysupgrade.bin
+	IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | append-metadata
+	DEVICE_PACKAGES := -kmod-ath10k-ct kmod-ath10k-ct-smallbuffers ipq-wifi-ezviz_cs-w3-wd1200g-eup
```
The two lines are longer than 80 characters. Please wrap them.

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -500,4 +500,20 @@ define Device/zyxel_wre6606
 endef
 TARGET_DEVICES += zyxel_wre6606
 
+define Device/ezviz_cs-w3-wd1200g-eup
```
Sorting.

adrianschmutzler commented on target/linux/ipq40xx/patches-4.19/901-arm-boot-add-dts-files.patch:
```diff
@@ -27,6 +27,7 @@ Signed-off-by: John Crispin <john@phrozen.org>
 +	qcom-ipq4018-nbg6617.dtb \
 +	qcom-ipq4018-rt-ac58u.dtb \
 +	qcom-ipq4018-wre6606.dtb \
++	qcom-ipq4018-cs-w3-wd1200g-eup.dtb \
```
Sorting.

Tom-Brouwer commented:
Hi @adrianschmutzler ,

Thanks for the review! In the latest commit I've addressed your comments w.r.t. sorting, whitespace & dts-license. Additionally, i've made minor changes to the WiFi MAC Adresses, and included 'label-mac-device' in the dts.

I've also implemented a solution in dts to reverse the LAN/WAN macs so it matches the factory firmware, but I'm not totally sure about it so please review.

There are multiple ways to reverse the LAN/WAN Macs:

1. I've already tried to switch the two eth interfaces by changing the 'ucidef_set_interfaces_lan_wan' and 'ucidef_add_switch' lines in 02_network. However, by default, eth1 seems to be hardwired to port 5 on my switch, ending up with LAN on the WAN port, so this doesn't work
2. Simply switch the eth1 and eth0 MAC addresses in the 'ipq40xx_setup_macs' function in 02_network. This will put the correct MAC addresses in the UCI network config.
3. Switch the gmac0 and gmac1 (eth0 and eth1) in the DTS file. This is the solution that comes closest to factory firmware (on factory firmware WAN=eth0 and LAN=eth1). I've extracted the [DTS from Factory firmware](https://github.com/openwrt/openwrt/files/4022144/extracted_ezviz_w3_dts.zip) and compared the 'gmac0' and 'gmac1' definition to the default DTS for IPQ4019 (see line 75 [here](https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/patches-4.19/711-dts-ipq4019-add-ethernet-essedma-node.patch#L75)), and found that in the default OpenWRT dts file, the gmac's are reversed compared to the factory dts. Therefore, I've changed these to the same configuration/order as in the factory DTS.

I've currently implemented option (3), but I can imagine that this has implications for future maintainability (e.g. if the default 'gmac0' and 'gmac1' are updated, because of driver changes). Is this an issue (or is it perhaps possible to inherit the gmac0/gmac1 from the included dts, but reverse their names?), or do you think the current solution is good?

Please let me know what you think, then I'll squash the commits. Thanks in advance!

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -257,6 +257,24 @@ define Device/engenius_ens620ext
+	SOC := qcom-ipq4018
+	IMAGES := sysupgrade.bin
+	IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs \
+		| append-metadata
```
Keep the pipe in the first line (that seems to be the most frequent wrapping scheme used):

	IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | \
		append-metadata

adrianschmutzler commented on target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh:
```diff
@@ -9,6 +9,12 @@ preinit_set_mac_address() {
 		;;
+	ezviz,cs-w3-wd1200g-eup)
+		wan_mac=$(mtd_get_mac_binary "ART" 0x0000)
+		lan_mac=$(mtd_get_mac_binary "ART" 0x0006)
```
I'd use 0x0 and 0x6 instead, but that's just optional.

adrianschmutzler commented:
Thanks for addressing my remarks. Note that I won't do a full review, so you will have to wait for someone else for that.

Tom-Brouwer commented:
Thanks for this second review @adrianschmutzler ! I've addressed your two remarks to clean the code a bit more. I've squashed the commits, and will wait for someone to do the full review. Thanks again!

adrianschmutzler commented on target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh:
```diff
@@ -9,6 +9,12 @@ preinit_set_mac_address() {
+		wan_mac=$(mtd_get_mac_binary "ART" 0x0)
+		lan_mac=$(mtd_get_mac_binary "ART" 0x6)
+		ip link set dev eth0 address $lan_mac
+		ip link set dev eth1 address $wan_mac
```
I'd make two lines out of these four.

adrianschmutzler commented:
> 2.	Simply switch the eth1 and eth0 MAC addresses in the 'ipq40xx_setup_macs' function in 02_network. This will put the correct MAC addresses in the UCI network config.
> 3.	Switch the gmac0 and gmac1 (eth0 and eth1) in the DTS file. This is the solution that comes closest to factory firmware (on factory firmware WAN=eth0 and LAN=eth1). I've extracted the DTS from Factory firmware and compared the 'gmac0' and 'gmac1' definition to the default DTS for IPQ4019 (see line 75 here), and found that in the default OpenWRT dts file, the gmac's are reversed compared to the factory dts. Therefore, I've changed these to the same configuration/order as in the factory DTS.

I can live with both variant 2 and 3. I personally care most about the MAC address assignment for LAN/WAN, and less about eth0/eth1, because the first ones will be what's visible/observable to most of the users. If I understand it correctly, both variant 2 and 3 will serve that purpose, so the users has the same address on LAN and WAN he would have with vendor firmware.

> I've currently implemented option (3), but I can imagine that this has implications for future maintainability (e.g. if the default 'gmac0' and 'gmac1' are updated, because of driver changes). Is this an issue (or is it perhaps possible to inherit the gmac0/gmac1 from the included dts, but reverse their names?), or do you think the current solution is good?

We had a similar situation in ath79 target, and if someone changes stuff like that, we will have to investigate anyway. The most important step to make this easier is to document your setup thoroughly in the commit message.

So, keep this for the moment and wait for someone to review this as a whole. :-)

Tom-Brouwer commented:
> I can live with both variant 2 and 3. I personally care most about the MAC address assignment for LAN/WAN, and less about eth0/eth1, because the first ones will be what's visible/observable to most of the users. If I understand it correctly, both variant 2 and 3 will serve that purpose, so the users has the same address on LAN and WAN he would have with vendor firmware.

I now have version 2 implemented, where label_mac is set in 02_network, and the WAN/LAN Mac are set in `preinit/05_set_iface_mac_ipq40xx.sh `, since this seems more in-line with other devices on the target.

I've also addressed the comment w.r.t. reducing the number of lines in that file. Thanks!

blocktrron commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -257,6 +257,24 @@ define Device/engenius_ens620ext
 TARGET_DEVICES += engenius_ens620ext
 
+define Device/ezviz_cs-w3-wd1200g-eup
+	$(call Device/FitImage)
```
Does the device support LZMA packed kernels? If so, an LZMA packed kernel saves quite some space on the SPI.

Tom-Brouwer replied:
Good question. I just tested this, and the shows "Unimplemented compression type 3" if I try to boot a LZMA image, so this does not seem to be supported.

blocktrron commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -257,6 +257,24 @@ define Device/engenius_ens620ext
+	DEVICE_VARIANT := EUP
+	BOARD_NAME := cs-w3-wd1200g-eup
+	DEVICE_DTS_CONFIG := config@4
+	KERNEL_SIZE := 4096k
```
Is this the kernel size from the vendor firmware? It's not used anywhere, so it could be dropped.

Tom-Brouwer replied:
Your are correct. I've removed the the KERNEL_SIZE variable

chunkeey commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-cs-w3-wd1200g-eup.dts:
```diff
@@ -0,0 +1,242 @@
+	status = "okay";
+	cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>;
+
+	mx25l12805d@0 {
```
If possible (u-boot needs to be happy with it) this node name should be a "generic term" like "flash@0". 
See the example on https://elinux.org/Device_Tree_Usage#Node_Names

"<name> is a simple ascii string and can be up to 31 characters in length. In general, nodes are named according to what kind of device it represents. ie. A node for a 3com Ethernet adapter would be use the name ethernet, not 3com509. "

Tom-Brouwer replied:
Thanks for the advice! I've changed it to flash@0, and tested it on the device, which works with this change.

chunkeey commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -26,6 +26,7 @@ endef
 ALLWIFIBOARDS:= \
 	aruba_ap-303 \
 	avm_fritzrepeater-1200 \
+	ezviz_cs-w3-wd1200g-eup \
```
I can't find the board-files in the ath10k mailing-list archive.

Please see the comment in ipq-wifi's Makefile if you need a pointer/template:

    # This is intended to be used on an interim basis until device-specific
    # board data for new devices is available through the upstream compilation
    #
    # Please send a mail with your device-specific board files upstream.
    # You can find instructions and examples on the linux-wireless wiki:
    # <https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles>

Tom-Brouwer replied:
I was under the impression that it would take some time for the boardfiles to be included in OpenWRT, so I was planning to submit them after the code was merged to master. Anyway, I now submitted the request to add the board-files (http://lists.infradead.org/pipermail/ath10k/2020-January/014620.html)

Tom-Brouwer committed:
ipq40xx: add support for EZVIZ CS-W3-WD1200G EUP

Hardware:
SOC:    Qualcomm IPQ4018
RAM:	128 MB Nanya NT5CC64M16GP-DI
FLASH:  16 MB Macronix MX25L12805D
ETH:    Qualcomm QCA8075 (4 Gigabit ports, 3xLAN, 1xWAN)
WLAN:   Qualcomm IPQ4018 (2.4 & 5 Ghz)
BUTTON: Shared WPS/Reset button
LED:    RGB Status/Power LED
SERIAL: Header J8 (UART, Left side of board). Numbered from
        top to bottom:
        (1) GND, (2) TX, (3) RX, (4) VCC (White triangle
        next to it).
        3.3v, 115200, 8N1

Tested/Working:
* Ethernet
* WiFi (2.4 and 5GHz)
* Status LED
* Reset Button (See note below)

Implementation notes:
* The shared WPS/Reset button is implemented as a Reset button
* I could not find a original firmware image to reverse engineer, meaning
currently it's not possible to flash OpenWRT through the Web GUI.

Installation (Through Serial console & TFTP):
1. Set your PC to fixed IP 192.168.1.12, Netmask 255.255.255.0, and connect to
one of the LAN ports
2. Rename the initramfs image to 'C0A8010B.img' and enable a TFTP server on
your pc, to serve the image
2. Connect to the router through serial (See connection properties above)
3. Hit a key during startup, to pause startup
4. type `setenv serverip 192.168.1.12`, to set the tftp server address
5. type `tftpboot`, to load the image from the laptop through tftp
6. type `bootm` to run the loaded image from memory
6. (If you want to return to stock firmware later, create an full MTD backup,
e.g. using instructions here https://openwrt.org/docs/guide-user/installation/generic.backup#create_full_mtd_backup)
7. Transfer the 'sysupgrade' OpenWRT firmware image from PC to router, e.g.:
`scp xxx-squashfs-sysupgrade.bin root@192.168.1.1:/tmp/upgrade.bin`
8. Run sysupgrade to permanently install OpenWRT to flash: `sysupgrade -n /tmp/upgrade.bin`

Revert to stock:
To revert to stock, you need the MTD backup from step 6 above:
1. Unpack the MTD backup archive
2. Transfer the 'firmware' partition image to the router (e.g. mtd8_firmware.backup)
3. On the router, do `mtd write mtd8_firmware.backup firmware`

Signed-off-by: Tom Brouwer <tombrouwer@outlook.com>

chunkeey commented:
Merged. Thanks.
